File: geogebra_bridge.js
/**
 * GeoGebra Bridge - 通过 Puppeteer/CDP 直接操控 GeoGebra Classic 6
 *
 * 用法:
 *   node geogebra_bridge.js <command_file.json>
 *   node geogebra_bridge.js --inline '["A = (0,0)", "B = (6,0)", "Segment(A,B)"]'
 */

const puppeteer = require('puppeteer-core');
const fs = require('fs');
const path = require('path');

class GeoGebraController {
  constructor() {
    this.browser = null;
    this.page = null;
  }

  async connect() {
    this.browser = await puppeteer.connect({
      browserURL: 'http://localhost:9222',
      defaultViewport: null
    });
    const pages = await this.browser.pages();
    this.page = pages.find(p => p.url().includes('classic')) || pages[0];
    const title = await this.page.title();
    console.log(`[GeoGebra] 已连接: ${title}`);
    return this;
  }

  async eval(jsCode) {
    return await this.page.evaluate(jsCode);
  }

  async cmd(command) {
    const result = await this.page.evaluate((cmd) => {
      try {
        return ggbApplet.evalCommand(cmd);
      } catch (e) {
        return 'Error: ' + e.message;
      }
    }, command);
    return result;
  }

  async cmdBatch(commands) {
    const results = [];
    for (const cmd of commands) {
      const r = await this.cmd(cmd);
      results.push({ command: cmd, result: r });
      if (r === true || r === 'true') {
        console.log(`  ✓ ${cmd}`);
      } else {
        console.log(`  ✗ ${cmd} → ${r}`);
      }
    }
    return results;
  }

  async resetView() {
    await this.eval(() => ggbApplet.reset());
  }

  async setPerspective(p) {
    await this.eval((persp) => ggbApplet.setPerspective(persp), p);
  }

  async getXML() {
    return await this.eval(() => ggbApplet.getXML());
  }

  async setXML(xml) {
    await this.eval((x) => ggbApplet.setXML(x), xml);
  }

  async getBase64() {
    return await this.eval(() => ggbApplet.getBase64());
  }

  async saveGGB(filepath) {
    const b64 = await this.getBase64();
    if (b64) {
      fs.writeFileSync(filepath, Buffer.from(b64, 'base64'));
      console.log(`[GeoGebra] 已保存: ${filepath}`);
      return true;
    }
    return false;
  }

  async exportPNG(filepath, scale = 2) {
    await this.eval((s) => ggbApplet.writePNGtoFile('__temp_export.png', s), scale);
    // writePNGtoFile saves to a fixed location; use getPNGBase64 instead
    const b64 = await this.eval((s) => ggbApplet.getPNGBase64(s), scale);
    if (b64) {
      fs.writeFileSync(filepath, Buffer.from(b64, 'base64'));
      console.log(`[GeoGebra] PNG 已保存: ${filepath}`);
      return true;
    }
    return false;
  }

  async newConstruction() {
    await this.eval(() => ggbApplet.newConstruction());
  }

  async disconnect() {
    if (this.browser) {
      await this.browser.disconnect();
      console.log('[GeoGebra] 已断开');
    }
  }
}

// ── 机构运动简图构建器 ──

class MechanismBuilder {
  constructor(ggb) {
    this.ggb = ggb;
  }

  /** 固定铰链 (机架铰点) */
  async fixedHinge(label, x, y) {
    await this.ggb.cmd(`${label} = (${x}, ${y})`);
    // 绘制固定铰链符号：小圆 + 填充
    const r = Math.max(0.15, Math.abs(Math.min(x, y) * 0.05 + 0.1));
    await this.ggb.cmd(`Circle(${label}, ${r.toFixed(3)})`);
    return label;
  }

  /** 活动铰链 (连杆铰点) */
  async revoluteJoint(label, x, y) {
    await this.ggb.cmd(`${label} = (${x}, ${y})`);
    return label;
  }

  /** 连杆 (连接两个铰链) */
  async link(label, pointA, pointB) {
    await this.ggb.cmd(`Segment(${pointA}, ${pointB})`);
    return label;
  }

  /** 曲柄 (绕固定铰链旋转的杆) */
  async crank(center, length, angleDeg = 0) {
    const rad = angleDeg * Math.PI / 180;
    const x = length * Math.cos(rad);
    const y = length * Math.sin(rad);
    const label = `CrankEnd`;
    await this.ggb.cmd(`${label} = ${center} + (${x.toFixed(4)}, ${y.toFixed(4)})`);
    await this.ggb.cmd(`Segment(${center}, ${label})`);
    // 给端点加圆标记
    await this.ggb.cmd(`Circle(${label}, 0.1)`);
    return label;
  }

  /** 滑块 (沿直线导轨移动) */
  async slider(label, railPoint1, railPoint2, initialRatio = 0.5) {
    await this.ggb.cmd(`${label}_rail = Segment(${railPoint1}, ${railPoint2})`);
    await this.ggb.cmd(`${label} = Point(${label}_rail, ${initialRatio})`);
    // 绘制滑块矩形符号
    return label;
  }

  /** 给点添加轨迹追踪 */
  async trace(pointLabel) {
    await this.ggb.eval((label) => ggbApplet.setTrace(label, true), pointLabel);
  }

  /** 设置点/线的颜色 */
  async setColor(label, r, g, b) {
    await this.ggb.eval(
      (l, cr, cg, cb) => ggbApplet.setColor(l, cr, cg, cb),
      label, r, g, b
    );
  }

  /** 设置线宽 */
  async setThickness(label, t) {
    await this.ggb.eval(
      (l, thickness) => ggbApplet.setLineThickness(l, thickness),
      label, t
    );
  }

  /** 隐藏标签 */
  async hideLabel(label) {
    await this.ggb.eval(
      (l) => ggbApplet.setLabelVisible(l, false),
      label
    );
  }
}

// ── CLI ──

async function main() {
  const args = process.argv.slice(2);
  if (args.length === 0 || args.includes('--help') || args.includes('-h')) {
    console.log(`GeoGebra Bridge - 通过 CDP 操控 GeoGebra Classic 6

用法:
  node geogebra_bridge.js <commands.json>  [--save output.ggb] [--png output.png]
  node geogebra_bridge.js --inline <json_array>
  node geogebra_bridge.js --xml <xml_string>
  node geogebra_bridge.js --base64

示例 JSON 文件格式:
{
  "perspective": "G",
  "commands": [
    "A = (0, 0)",
    "B = (6, 0)",
    "C = (2, 4)",
    "Segment(A, B)",
    "Segment(B, C)",
    "Segment(C, A)"
  ],
  "save": "output.ggb",
  "png": "output.png"
}
`);
    process.exit(0);
  }

  const ggb = new GeoGebraController();
  await ggb.connect();

  try {
    let savePath = null;
    let pngPath = null;

    if (args.includes('--inline')) {
      const idx = args.indexOf('--inline');
      const cmds = JSON.parse(args[idx + 1]);
      ggb.newConstruction();
      await ggb.cmdBatch(cmds);
      ggb.resetView();

    } else if (args.includes('--xml')) {
      const idx = args.indexOf('--xml');
      const xml = args[idx + 1];
      await ggb.setXML(xml);

    } else {
      // 从文件加载
      const filepath = args[0];
      const config = JSON.parse(fs.readFileSync(filepath, 'utf-8'));

      if (config.perspective) {
        await ggb.setPerspective(config.perspective);
      }

      if (config.new_construction !== false) {
        await ggb.newConstruction();
      }

      if (config.commands && config.commands.length > 0) {
        console.log(`[GeoGebra] 执行 ${config.commands.length} 条命令:`);
        await ggb.cmdBatch(config.commands);
      }

      if (config.xml) {
        await ggb.setXML(config.xml);
      }

      await ggb.resetView();

      savePath = config.save || config.output;
      pngPath = config.png || config.screenshot;
    }

    // 处理 --save 和 --png 参数覆盖
    const saveIdx = args.indexOf('--save');
    if (saveIdx >= 0) savePath = args[saveIdx + 1];
    const pngIdx = args.indexOf('--png');
    if (pngIdx >= 0) pngPath = args[pngIdx + 1];

    if (savePath) {
      await ggb.saveGGB(savePath);
    }
    if (pngPath) {
      await ggb.exportPNG(pngPath);
    }

    if (args.includes('--base64')) {
      const b64 = await ggb.getBase64();
      console.log(b64);
    }

  } finally {
    await ggb.disconnect();
  }
}

main().catch(err => {
  console.error('错误:', err.message);
  process.exit(1);
});

module.exports = { GeoGebraController, MechanismBuilder };

File: geogebra_bridge.py
"""
GeoGebra CDP Bridge - 通过 Chrome DevTools Protocol 直接操控 GeoGebra Classic 6
"""
import json
import time
import websocket

class GeoGebraBridge:
    """通过 CDP WebSocket 连接 GeoGebra，执行 JavaScript 控制几何构造"""

    def __init__(self, host="localhost", port=9222):
        self.host = host
        self.port = port
        self.ws = None
        self._msg_id = 0
        self._pending = {}

    def connect(self):
        """连接到 GeoGebra 的 CDP 端点"""
        # 获取可调试页面列表
        import urllib.request
        resp = urllib.request.urlopen(f"http://{self.host}:{self.port}/json")
        pages = json.loads(resp.read())

        # 找到 GeoGebra 经典页面
        geo_page = None
        for p in pages:
            if "GeoGebra" in p.get("title", "") or "classic" in p.get("url", ""):
                geo_page = p
                break
        if not geo_page and pages:
            geo_page = pages[0]

        if not geo_page:
            raise RuntimeError("没有找到 GeoGebra 页面")

        ws_url = geo_page["webSocketDebuggerUrl"]
        print(f"[Bridge] 连接到: {geo_page.get('title', 'GeoGebra')}")

        self.ws = websocket.create_connection(ws_url, header={
            "Origin": "http://127.0.0.1:9222",
            "User-Agent": "GeoGebraBridge/1.0"
        })
        self.ws.settimeout(10)
        return self

    def _send(self, method, params=None):
        self._msg_id += 1
        msg = {
            "id": self._msg_id,
            "method": method,
            "params": params or {}
        }
        self.ws.send(json.dumps(msg))
        # 等待响应
        while True:
            resp = json.loads(self.ws.recv())
            if resp.get("id") == self._msg_id:
                if "error" in resp:
                    raise RuntimeError(f"CDP Error: {resp['error']}")
                return resp.get("result", {})
            # 可能是事件消息，继续等待

    def eval(self, js_code, timeout=10):
        """在 GeoGebra 上下文中执行 JavaScript 代码"""
        result = self._send("Runtime.evaluate", {
            "expression": js_code,
            "returnByValue": True,
            "timeout": timeout * 1000
        })
        return result

    def ggb_command(self, cmd):
        """执行 GeoGebra 命令（如创建点、线等）"""
        result = self.eval(f"ggbApplet.evalCommand({json.dumps(cmd)})")
        val = result.get("result", {}).get("value", None)
        return val

    def ggb_get_xml(self):
        """获取当前构造的 XML"""
        result = self.eval("ggbApplet.getXML()")
        return result.get("result", {}).get("value", "")

    def ggb_get_base64(self):
        """获取当前构造的 GGB 文件 (base64)"""
        result = self.eval("ggbApplet.getBase64()")
        return result.get("result", {}).get("value", "")

    def ggb_save(self, filepath):
        """保存当前构造为 .ggb 文件"""
        import base64
        b64 = self.ggb_get_base64()
        if b64:
            with open(filepath, "wb") as f:
                f.write(base64.b64decode(b64))
            print(f"[Bridge] 已保存: {filepath}")
            return True
        return False

    def ggb_set_perspective(self, perspective="G"):
        """设置视图: G=几何, A=代数, T=表格, 3D=三维"""
        self.eval(f"ggbApplet.setPerspective('{perspective}')")

    def ggb_new(self):
        """新建空白构造"""
        self.eval("ggbApplet.newConstruction()")

    def ggb_zoom_all(self):
        """缩放以适应所有对象"""
        self.eval("ggbApplet.zoomAll()")

    def ggb_export_png(self, filepath, scale=2):
        """导出为 PNG 图片"""
        import base64
        result = self.eval(f"ggbApplet.exportPNG(1, false, {scale})")
        data = result.get("result", {}).get("value", "")
        if data:
            with open(filepath, "wb") as f:
                f.write(base64.b64decode(data.split(",")[1] if "," in data else data))
            print(f"[Bridge] PNG 已保存: {filepath}")
            return True
        return False

    def close(self):
        if self.ws:
            self.ws.close()


# ── 便捷函数 ──

def create_mechanism(name, commands, output_dir="D:/tool"):
    """
    用 GeoGebra 命令列表创建机构运动简图并保存为 .ggb 和 PNG

    参数:
        name: 机构名称
        commands: GeoGebra 命令列表，如 ["A = (0, 0)", "B = (4, 0)", ...]
        output_dir: 输出目录
    """
    import os
    ggb = GeoGebraBridge().connect()
    ggb.ggb_new()

    for cmd in commands:
        result = ggb.ggb_command(cmd)
        if result is not None:
            print(f"  ✓ {cmd} → {result}")
        else:
            print(f"  ✗ {cmd} (可能失败)")

    ggb.ggb_zoom_all()
    time.sleep(0.5)

    ggb_path = os.path.join(output_dir, f"{name}.ggb")
    png_path = os.path.join(output_dir, f"{name}.png")
    ggb.ggb_save(ggb_path)
    ggb.ggb_export_png(png_path)
    ggb.close()
    return ggb_path, png_path
